### PR TITLE
Fix RIDs for the native dependencies in Grpc.Core

### PR DESCRIPTION
**Pull request by EraYaN**
I couldn't really find any template to generate this project file so I hope this can be hand edited.

This should fix the issues mentioned in #18106 and address one of the comments from #17451

This should work also work for musl based linux versions (as you can see in the file linked below `linux-musl-x64` includes `linux-x64`).
Technically `osx-x86` doesn't exist ([Reference](https://github.com/dotnet/runtime/blob/master/src/libraries/pkg/Microsoft.NETCore.Platforms/runtime.json))

I'm not sure what to do for the android ones, technically they only seems to support `android-arm`, `andoird-arm64`, `android-x64` and `android-x86` and I'm not sure how the nested folders get picked up.

In the thread linked above, there was also a mention of splitting the native assets from Grpc.Core. As an approach you can take the route SkiaSharp took.

```
Grpc.Core -> Grpc.Core.NativeAssets.Linux
          -> Grpc.Core.NativeAssets.Windows
          -> Grpc.Core.NativeAssets.MacOS
```
With separate packages for like `Grpc.Core.NativeAssets.Linux.Arm`, `Grpc.Core.NativeAssets.Android` etc. Although that should maybe be a separate PR.

@yashykt

### CODE REVIEW

**Comment by EraYaN** [comment]
The commit ID issue should be rectified now.

**Comment by jtattermusch** [comment]
Adhoc run of distribtests: sponge/ccfd0af3-5be0-42fd-9f45-dd02b4b11787 (will post results later)

**Comment by EraYaN** [comment]
Seemed like master removed the x86 versions for macos and linux, I have updated this PR with those changes.

File: src/csharp/Grpc.Core/Grpc.Core.csproj [hunk 1]
```diff
@@ -36,19 +36,19 @@
   <ItemGroup>
     <EmbeddedResource Include="..\..\..\etc\roots.pem" />
     <Content Include="..\nativelibs\csharp_ext_macos_x64\libgrpc_csharp_ext.dylib">
-      <PackagePath>runtimes/osx/native/libgrpc_csharp_ext.x64.dylib</PackagePath>
+      <PackagePath>runtimes/osx-x64/native/libgrpc_csharp_ext.x64.dylib</PackagePath>
```
Comment (jtattermusch) [inline src/csharp/Grpc.Core/Grpc.Core.csproj:39]: Changing this file is not enough, as there is other logic that relies on knowing these exact paths.

E.g. 
https://github.com/grpc/grpc/blob/9936d7785cdf41c13974497423c7d87bfbbd4054/src/csharp/Grpc.Core/build/net45/Grpc.Core.targets (which is required for running on legacy .NET frameworks other than Core CLR).

and
https://github.com/grpc/grpc/blob/9936d7785cdf41c13974497423c7d87bfbbd4054/src/csharp/Grpc.Core/Internal/NativeExtension.cs#L83

**Review by jtattermusch — CHANGES_REQUESTED** [review]
Thanks for the contribution, but there's several problems with this PR (there are multiple other locations that would need to be updated, otherwise things would be broken).

File: src/csharp/Grpc.Core/Grpc.Core.csproj (hunk 1 above)
Comment (EraYaN) [reply src/csharp/Grpc.Core/Grpc.Core.csproj:39]: Alright I have addressed these issues I think. For the core platforms (and xamarin too) the project should probably rely in the "smart" SDK for finding the correct native binary instead of creating the list itself. The runtime seems to be able handle the search paths (with all the fallback RIDs etc) and native library extensions (.so, .dll, dylib) by itself. I'm not sure how to access that logic with the way the native symbols are loaded currently, but it might be worth exploring.

**Comment by jtattermusch** [comment]
Triggered adhoc distribtest run: 
https://g3c.corp.google.com/results/invocations/0e0df023-3870-4cb3-b005-32ee9b9194a4/targets (will post back results later).

**Comment by EraYaN** [comment]
The tests were failing because I made yet another mistake by not updating the project files correctly, and locally it was working because I had cached copies of the binaries.

File: src/csharp/Grpc.Core/Grpc.Core.csproj [hunk 2]
```diff
@@ -6,7 +6,6 @@
     <Authors>The gRPC Authors</Authors>
     <Copyright>Copyright 2015 The gRPC Authors</Copyright>
     <Description>C# implementation of gRPC based on native gRPC C-core library.</Description>
-    <PackageIconUrl>https://github.com/grpc/grpc.github.io/raw/master/img/grpc_square_reverse_4x.png</PackageIconUrl>
```
Comment (jtattermusch) [inline src/csharp/Grpc.Core/Grpc.Core.csproj:9]: let's not do unrelated changes like this in this PR.

File: src/csharp/Grpc.Core/build/net45/Grpc.Core.targets [hunk 3]
```diff
@@ -10,22 +10,22 @@
     the Grpc.Core.dll assembly and the native libraries are used at all times.
   -->
   <ItemGroup Condition="'$(Grpc_SkipNativeLibsCopy)' != 'true'">
-    <Content Include="$(MSBuildThisFileDirectory)..\..\runtimes\win\native\grpc_csharp_ext.x86.dll">
+    <Content Include="$(MSBuildThisFileDirectory)..\..\runtimes\win-x86\native\grpc_csharp_ext.dll">
       <CopyToOutputDirectory>PreserveNewest</CopyToOutputDirectory>
       <Link>grpc_csharp_ext.x86.dll</Link>
```
Comment (jtattermusch) [inline src/csharp/Grpc.Core/build/net45/Grpc.Core.targets:15]: wouldn't the <Link> need to be updated as well?

Comment (jtattermusch) [reply src/csharp/Grpc.Core/build/net45/Grpc.Core.targets:15]: the problem is that the NativeExtension would need to be able to find the file that's copied to the output folder  on net45 (but you just changed its logic to look for grpc_csharp_ext.dll only.

At the same time, there are 2 `<Content>` items: one for x86 and one for x64.
you cannot copy them both into a single grpc_csharp_ext.dll (I think this is the original reason why we have the x64.dll and x86.dll suffix - to avoid a name clash in net45.

**Comment by jtattermusch** [comment]
Another run of distribtests: https://g3c.corp.google.com/results/invocations/9d78c3bd-04a9-4b0c-befb-796f6f290037/targets (will post results later, but I'm already pretty sure there are still bugs).

**Review by jtattermusch — CHANGES_REQUESTED** [review]
I think there's still issues with the PR as is.

**Comment by jtattermusch** [comment]
I took what what in this PR and made a few adjustments so that things work correctly (I believe that's easier that back and forth on this PR) https://github.com/grpc/grpc/pull/24766 - feel free to comment there.
